![AIMS Logo](https://www.uni-bielefeld.de/fakultaeten/mathematik/ag/hofmanova/daad-aims/aims-senegal.png)

****

# <center> <b> <span style="color:orange;"> Python Proficiency for Scientific Computing and Data Science (PyPro-SCiDaS)  </span> </b></center>

### <center> <b> <span style="color:green;">An Initiation to Programming using Python (Init2Py) </span> </b></center>
    

****
# <center> <b> <span style="color:blue;">Assignment </span> </b></center>

****
### Instructions:
You should solve these problems using only:
- Standard Python data types (e.g., str, list, dict, int).
- Built-in and user-defined functions.
- Basic flow control (if/else, for loops, while loops).
- Standard library modules such as math
****

## Section 1: Python Fundamentals

### Exercise 1
Compute the following expressions for a natural number $n$ chosen by the user:

a) $$\sum_{\substack{i=0 \\ i \text{ is odd}}}^{n-1} (i + 3)$$


In [1]:
#Your code here
def somme_impairs(n):
    total = 0
    for i in range(1, n, 2):
        total += i + 3
    return total

n = int(input("entrez un numbre "))
resultat = somme_impairs(n)
print(f"Le résultat pour n = {n} est : {resultat}")        

entrez un numbre  12


Le résultat pour n = 12 est : 54


b) $$\prod_{\substack{i=0 \\ i \text{ is even}}}^{n-1} (i + 3)$$

In [ ]:
# Your code here
def compute_even_product(n):
    product = 1
    for i in range(n):
        if i % 2 == 0:  # I try to check if i is even
            product *= (i + 3) # I multiply product by (i + 3)
    return product
n = 3
result = compute_even_product(n)
print(f"The result for n = {n} is: {result}")


The result for n = 3 is: 15


### Exercise 2
a) Write python code that computes the sum of the square of the first $n$ natural numbers. One version with a "for loop" and another version with a "while loop".

In [ ]:
#Part a: For loop version
def somme_carres_for(n):
    total = 0
    for i in range(1, n + 1): 
        total += i**2
    return total

n = int(input("entrez une valeur pour n :"))

resultat_for = somme_carres_for(n)

print(f"La somme des carrés des {n} premiers entiers (avec for) est : {resultat_for}")

In [ ]:
#Part b: While loop version
def somme_carres_while(n):
    somme = 0
    i = 1
    while i <= n:
        somme += i**2
        i += 1
    return somme
n = int(input("Entrez une valeur de n "))
resultat_while = somme_carres_while(n)
print(f"computes the sum of the square of the first{n} is : {resultat_while}")

Entrez une valeur de n 10
La somme des carrés des 10 premiers entiers (avec while) est : 385


b) Write python code that prints, for a natural number $M$, the smallest natural number $n$ such that $1^2 + 2^2 + \cdots + n^2 \geq M$. One version with a "for loop" and another version with a "while loop".

In [ ]:
#Part b: While loop version
M = int(input("Entrez un nombre naturel M : "))
somme = 0
n = 0
while somme < M:
    n += 1
    somme += n ** 2

print("The smallest n is :", n)

The smallest  n is : 3


In [ ]:
#Part b: For loop version
M = int(input("Entrez un nombre naturel M : "))
somme = 0
for n in range(1, M + 1):  # M+1 est arbitraire, on sort avec break
    somme += n ** 2
    if somme >= M:
        print("Le plus petit n est :", n)
        break

### Exercise 3
Write a python script that asks the user for the values of $x$ and $y$ and uses these inputs to compute the numerical value of the expression:

$$\frac{\sqrt{x^2+y}}{\frac{1}{x} + 2\sin y}$$

In [ ]:
#Your code here
import math
x=int(input("Enter the value of x :"))
y=int(input("Enter the value of y  :"))
y_rad = math.radians(y) # I convert y to radians for the sine function

try:
    numerator = math.sqrt( (x**2)+y)
    denominator = ((1/x)+ 2*math.sin(y_rad))
except ZeroDivisionError:
    print("Error: Division by zero.!")
else:
    result = numerator / denominator
    print(f"The result of the expression is: {result}")

print (" La reponse final est : " , result)


The result of the expression is: 2.7838821814150108
 La reponse final est :  2.7838821814150108


### Exercise 4

A store charges XOF $12$ per item if fewer than $10$ items are purchased. If the number of items is between $10$ and $99$ (inclusive), the price is XOF $10$ per item. For $100$ items or more, the price drops to XOF $7$ per item.

a) Write a function called `Cost` that takes as input the number of items $n$ and returns the total cost.


In [64]:
#Your code here
def cost(number_item): 
    total_cost = 0
    if number_item < 10:
        total_cost = 12*number_item
    elif number_item in range(10,100):
        total_cost = 10*number_item
    if number_item >= 100:
        total_cost = 7*number_item
    return total_cost

try:
    nombre_item = int(input("Enter a nomber of items : "))
    total_final = cost(nombre_item)
    if nombre_item < 0:
        raise ValueError("The number of items cannot be negative.")
except TypeError:
    print(" Error: Invalid input type. Please enter an integer value.")
else:
    print("The total cost of items is :", total_final)

The total cost of items is : 700


b) Test the function `Cost` for $n = 50 \text{ and } 255$.

In [57]:
#Your code here
print("pour n = 5, le total :", cost(5))
print("pour n = 255, le total : ", cost(255))

pour n = 5, le total : 60
pour n = 255, le total :  1785


### Exercise 5
For a fixed real number $x$ and a natural number $n$, we can define recursively $x^n$ using the relations:

$$x^0 = 1 \quad \text{and} \quad x^{n+1} = x \cdot x^n$$

Write a function `power(x,n)` that implements the above recursion.

In [34]:
#Your code here
def power(x, n):
    if n == 0:
        return 1
    else:
        return x * power(x, n - 1)

print(power(10,2))

100


### Exercise 6

a) Write two functions: `arithmetic_mean`, which takes two numbers and returns their [**arithmetic mean**](https://en.wikipedia.org/wiki/Arithmetic_mean)
, and `geometric_mean`, which takes two numbers and returns their [**geometric mean**](https://en.wikipedia.org/wiki/Geometric_mean). Then write a function `avggeom` that takes two input numbers and, for 10 iterations, repeatedly computes both their arithmetic mean and geometric mean, using the results from the previous iteration as the new inputs.


In [ ]:
#Your code here
def arithmetic_mean(a, b):
    return (a + b) / 2

def geometric_mean(a, b):
    return (a * b) ** 0.5

def avggeom(a, b):
    for _ in range(10):
        a_new = arithmetic_mean(a, b)
        b_new = geometric_mean(a, b)
        a, b = a_new, b_new
    return a, b

x = float(input("Enter the value of x : "))
y = float(input("Enter the value of y : "))
moy_arith, moy_geom = avggeom(x, y)
print(f"arithmetic mean : {moy_arith}, geometric mean : {moy_geom}")

result = avggeom(10, 4)
print(result)

Moyenne arithmétique : 3.3289971432731074, Moyenne géométrique : 3.3289971432731074
(6.657994286546215, 6.657994286546215)


b) Write a function called `distancefn` that computes the [**Euclidean distance**](https://en.wikipedia.org/wiki/Euclidean_distance) between two points $(x_1, y_1)$ and $(x_2, y_2)$ in the plane. Use it in another program that computes the perimeter and the area of a triangle, given three points in the plane (each given as coordinates).

 


In [ ]:
# Your code here
def distancefn(x1, y1, x2, y2):
    return ((x2 - x1)**2 + (y2 - y1)**2)**0.5
point1 = (2, 3)
point2 = (5, 7)
dist = distancefn(point1[0], point1[1], point2[0], point2[1])
print(f"La distance entre {point1} et {point2} est : {dist}")


La distance entre (2, 3) et (5, 7) est : 5.0


### Exercise 7

a) Let `L=[2,3,5,7,11,13,17,19,23,29,31,37,41,43,47]`. Use list comprehension to produce a list of the gaps (differences) between consecutive entries in `L`. Then find the maximum gap size and the percentage of gaps that have size $2$.

In [58]:
# Your code here
L=[2,3,5,7,11,13,17,19,23,29,31,37,41,43,47]
list_gap = [ L[i+1] - L[i] for i in range(len(L) - 1)]
print("Les écarts entre les nombres premiers consécutifs sont :", list_gap)

Les écarts entre les nombres premiers consécutifs sont : [1, 2, 2, 4, 2, 4, 2, 4, 6, 2, 6, 4, 2, 4]


b) i) Using list comprehensions, write a function `ListComp` that takes a list of strings as input and returns three objects:

- a list that consists of the strings with their first characters removed
- a list of the lengths of the strings
- a list that consists of only those strings that are at least 5 characters long


In [ ]:
# Your code here
def ListComp(strings):
    list1 = [s[1:] for s in strings if len(s) > 0]
    lengths = [len(s) for s in strings]
    list2 = [s for s in strings if len(s) >= 5]
    return list1, lengths, list2


ii) Test your function with the list: `["hello", "world", "python", "programming", "AI", "code"]`

In [31]:
# Your code here
list2 = ["hello", "world", "python", "programming", "AI", "code"]
print(ListComp(list2))

(['ello', 'orld', 'ython', 'rogramming', 'I', 'ode'], [5, 5, 6, 11, 2, 4], ['hello', 'world', 'python', 'programming'])


### Exercise 8

a) Write a program that repeatedly asks the user to enter product names and prices. Store all of these in a dictionary whose keys are the product names and whose values are the prices. When the user is done entering products and prices, allow them to repeatedly enter a product name and print the corresponding price or a message if the product is not in the dictionary.

In [ ]:
# Your code here
dictProducts={}
while True:
    try:
        product = input("Enter a name of product (or 'stop' for exit ) : ")
        if product== "stop":
            break
        else :
            price = float(input(f"Enter a price for {product} : " ))
            dictProducts[product] = price 
    except ValueError:
        print("Invalid input. Please enter a valid price.")
    
print(dictProducts)

def rechercheProduit(product_search, dict_input):
    trouve=False
    for s, value in dict_input.items() : 
        if s==product_search : 
            print(f" find product: {s}, price : {value}")
            trouve=True
            break
    if trouve== False:
        print("we've not found any product" )
    
search_value = input("Enter a product for searching ")
rechercheProduit(search_value, dictProducts)        

Invalid input. Please enter a valid price.
{'apple': 12.0, 'mangoes': 45.0, 'annanas': 15.0}
 find product: apple, price : 12.0


b) Using the dictionary created above, allow the user to enter an amount and print out a list of the products whose price is less than that amount.

In [ ]:
# Your code here
try : 
    montant = float(input("Entrez un montant pour filtrer les produits : "))
    print(f"Produits dont le prix est inférieur à {montant} :")
except ValueError:
    print("Invalid input. Please enter a valid amount.")
else :
    for nom, prix in dictProducts.items():
        if prix < montant:
            print(f"- {nom} : {prix}")

Produits dont le prix est inférieur à 12.0 :


c) Using the dictionary created above, evaluate the most expensive product.

In [103]:
# Your code here
try:
    montant = float(input("Entrez un montant pour filtrer les produits : "))
except ValueError:
    print("Invalid input. Please enter a valid amount.")
else:
    print(f"Produits dont le prix est inférieur à {montant} :")
    for nom, prix in dictProducts.items():
        if prix > montant:
            print(f"- {nom} : {prix}")

Produits dont le prix est inférieur à 12.0 :
- mangoes : 45.0
- annanas : 15.0


d) Using the dictionary created above, calculate the average price of all products.

In [ ]:
# Your code here
def average_price(dict_input):
    total = sum(dict_input.values())
    count = len(dict_input)
    if count == 0:
        return 0
    return total / count 

average = average_price(dictProducts)
print(f"Le prix moyen des produits est : {average}")

e) Using the dictionary created above, update the prices of multiple products at once.

In [ ]:
# Your code here
def update_price(name, new_price, dict_input):
    if name in dict_input:
        dict_input[name] = new_price
        print(f"The price of {name} has been updated {new_price}.")
    else:
        print(f"Product {name} not found.")   

try:
    product_to_update = input("Enter name of a product that updates : ")
except ValueError: # I want to handle invalid input
    print("Invalid input. Please enter a valid product name.")
else :
    new_price = float(input(f"Enter price for product {product_to_update} : "))
    update_price(product_to_update, new_price, dictProducts)


### Exercise 9

Use string methods and attributes to solve the following:

a) Write a function called `word_counter` that reads in the file `example.txt` and estimates the total number of words in the file. Assume words are separated by spaces.

In [2]:
# Your code here
def word_counter(file_path):
    
    try:
        with open(file_path, "r") as file:
            content = file.read()
            words = content.split()  
            return len(words) 
    except FileNotFoundError:
        print("file 'example.txt' no such files or directory.")
        return 0

total_words = word_counter("files/example.txt")
print("Nombre de mots dans le fichier :", total_words)


Nombre de mots dans le fichier : 6


b) Write a function called `UpLow` that reads the content of `example.txt` and then returns a dictionary containing the total number of uppercase letters and the total number of lowercase letters.

In [ ]:
# Your code here


c) Write a function called `count_word` that reads in `example.txt` and returns how many times the `target_word` ("the") appears in the text. Take note that the count must be case-insensitive (e.g., "The", "the", "THE" all count as one) and that it must only count whole word matches, not partial matches (e.g., "them" should not count if the target is "the").

In [ ]:
# Your code here


d) Write a function called `swap_and_clean` that reads the content of the file `example.txt`. The function should process the text by converting all uppercase letters to lowercase and all lowercase letters to uppercase. It should also remove all characters that are not letters, such as numbers and punctuation, while keeping spaces intact. After performing these modifications, the function should write the resulting text to a new file called `output1.txt`.

In [ ]:
# Your code here


e) Write a function called `palindromes` that reads the content of `example.txt`. The function should first split the text into individual words. For each word, it should be cleaned by converting it to lowercase and removing any leading or trailing punctuation. After cleaning, the function should check whether the word is a [**palindrome**](https://en.wikipedia.org/wiki/Palindrome). A palindrome is a word that reads the same forwards and backwards eg "madam". If a word is a palindrome, the function should print the original, uncleaned version of that word to the console.

In [ ]:
# Your code here


f) Write a function called `extract_and_format_numbers` that reads the file `example.txt` line by line. The function should identify any numbers that appear in the text, such as `100000000`. For each number found, it should format the number by inserting commas in the appropriate places (for example, converting `100000000` to `100,000,000`). Each formatted number should then be written on a separate line in a new file called `output2.txt`.

In [ ]:
# Your code here


False


### Exercise 10

The [**digital root**](https://en.wikipedia.org/wiki/Digital_root) of a positive number $n$ is obtained as follows: 
- Add up the digits $n$ to get a new number.
- Add up the digits of that to get another new number. Keep doing this until you get a number that has only one digit. That number is the digital root.

For example, if $n = 45893$, we add up the digits to get $$4 + 5 + 8 + 9 + 3 = 29.$$ We then add up
the digits of $29$ to get $$2 + 9 = 11.$$ We then add up the digits of $11$ to get $$1 + 1 = 2.$$ Since 2 has
only one digit, 2 is our digital root.

Write the Python code of the function `DigRoot` that returns the digital root of an integer $n$ given as input.



In [ ]:
# Your code here
def DigRoot(n):
    while n >= 10:
        digits = [int(d) for d in str(n)]
        total = sum(digits)
        print(" + ".join(str(d) for d in digits), "=", total)
        n = total
    return n

print(DigRoot(45893))  
print(DigRoot(123456)) 
print(DigRoot(9))


4 + 5 + 8 + 9 + 3 = 29
2 + 9 = 11
1 + 1 = 2
2


## Section 2: Mathematical Applications

### Exercise 1
Compute the first 30 terms of the sequence defined by $2x_{n+2} - x_{n+1} - 6x_n = 0$ sequence, given any two initial values $x_0$ and $x_1$.

In [ ]:
# Your code here
def sequence_x(x0, x1, terms=30):
    seq = [x0, x1]
    for n in range(2, terms):
        xn_plus_2 = 0.5 * (seq[n-1] + 6 * seq[n-2])
        seq.append(xn_plus_2)
    return seq
x0 = 1
x1 = 2
result = sequence_x(x0, x1)
print(result)

[1, 2, 4.0, 8.0, 16.0, 32.0, 64.0, 128.0, 256.0, 512.0, 1024.0, 2048.0, 4096.0, 8192.0, 16384.0, 32768.0, 65536.0, 131072.0, 262144.0, 524288.0, 1048576.0, 2097152.0, 4194304.0, 8388608.0, 16777216.0, 33554432.0, 67108864.0, 134217728.0, 268435456.0, 536870912.0]


### Exercise 2

Let $f$ be a continuous function defined on the closed interval $[a, b]$ such that $f(a)$ and $f(b)$ have opposite signs. By the [**Intermediate Value Theorem**](https://en.wikipedia.org/wiki/Intermediate_value_theorem), there exists a root $x_0 \in (a, b)$ for which $f(x_0) = 0$.

The [**bisection method**](https://fiveable.me/numerical-analysis-i/unit-3/bisection-method-algorithm/study-guide/5LfBknp8yfSe5lse) is an iterative method for approximating the root of such a function. At each step, it divides the interval in half and selects the subinterval where the sign change occurs, repeating this process until the interval width is smaller than a given tolerance ($\epsilon$).

a) Write a function named `bisection(f, a, b, epsilon=1e-2)` that:

- Returns a tuple $(\bar{x}, f(\bar{x}))$, where $\bar{x}$ is an approximation of the root such that
$$
|\bar{x} - x_0| \le \epsilon
$$
if $f(a)\cdot f(b) < 0$.  

- Prints ``Bisection method failed'' if $f(a)\cdot f(b) \ge 0$.


In [ ]:
# Your code here
#Bisection method with maximum iterations (bisection method shown by the algorithm in that previous link)
def bisection(f, a, b, tol, max_iter):
    if f(a) * f(b) >= 0:
        return "Error: No root in interval"

    for i in range(1, max_iter + 1):
        c = (a + b) / 2
        if abs(b - a) < tol or abs(f(c)) < tol:
            return c
        if f(c) * f(a) < 0:
            b = c
        else:
            a = c
    return "Error: Maximum iterations reached"


In [ ]:
def f(x):
    return x**2 - 2
#case with max_iter=100
root = bisection(f, 0, 2, tol=1e-6, max_iter=100)
print("Racine trouvée :", root)

Racine trouvée : 1.4142136573791504


In [ ]:
def bisection_v2(f, a, b, epsilon=1e-2):
    # Vérifions que f(a) et f(b) ont des signes opposés
    if f(a) * f(b) >= 0:
        print("Bisection method failed")
        return

    while abs(b - a) > epsilon:
        c = (a + b) / 2
        fc = f(c)
        if abs(fc) < epsilon:
            return (c, fc)
        if f(a) * fc < 0:
            b = c
        else:
            a = c
    c = (a + b) / 2
    return (c, f(c))

In [ ]:
# Exemple : f(x) = x^2 - 2, racine attendue ≈ √2
def f(x):
    return x**2 - 2

result = bisection_v2(f, 0, 2, epsilon=1e-4)
print("Approximation de la racine :", result)

Approximation de la racine : (1.4141845703125, -8.200109004974365e-05)


b) Test your function `bisection` for $ f(x) = x^3 -3x - 5$, $a=-10$ and $b=10$

In [45]:
# Your code here
def f(x):
    return x**3 - 3*x - 5

# Test avec a = -10, b = 10
result = bisection_v2(f, -10, 10)
print("Approximation de la racine :", result)

Approximation de la racine : (2.2802734375, 0.015796511434018612)


### Exercise 3
An irrational decimal fraction, [**Champernowne constant**](https://en.wikipedia.org/wiki/Champernowne_constant), is created by concatenating the positive integers: 0.12345678910<font color='red'>1</font>112131415161718192021...

It can be seen that the 12th digit of the fractional part is <font color='red'>1</font>. If $d_n$ represents the nth digit of the fractional part, find the value of the following expression.

$$d_1 × d_{10} × d_{100} × d_{1000} × d_{10000} × d_{100000} × d_{1000000}$$

In [ ]:
def champernowne_constant(n_digits):
    digits = ''
    n = 1
    while len(digits) < n_digits:
        digits += str(n)
        n += 1
    return '0.' + digits[:n_digits]

champernowne_constant(1) # I test first 1 digits
champernowne_constant(10) # I test first 10 digits
champernowne_constant(100) # I test first 100 digits
champernowne_constant(1000)

'0.1234567891'

In [ ]:
# Your code here
def champernowne_digit_at(n):
    digits = ''
    i = 1
    while len(digits) < n:
        digits += str(i)
        i += 1
    return int(digits[n - 1])

positions = [1, 10, 100, 1000, 10000, 100000, 1000000]
product = 1

for pos in positions:
    digit = champernowne_digit_at(pos)
    print(f"d [{pos}] = {digit}")
    product *= digit

print(f"\nProduit final : {product}")


d [1] = 1
d [10] = 1
d [100] = 5
d [1000] = 3
d [10000] = 7
d [100000] = 2
d [1000000] = 1

Produit final : 210


## Section 3: Advanced Programming

### Exercise 1

a) Write a Python function named `factorial2` that takes a non-negative integer $n$ as input and returns $n!$ (the factorial of $n$).

In [47]:
# Your code here
def factorial2(n):
    if n < 0:
        raise ValueError("Input must be a non-negative integer.")
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result

def recursive_factorial2(n):
    if n < 0:
        raise ValueError("Input must be a non-negative integer.")
    if n == 0 or n == 1:
        return 1
    else:
        return n * recursive_factorial2(n - 1)
print(factorial2(5)) 
print(recursive_factorial2(5))

120
120


b) i) Write a Python function named `QuadEqu` that solves a quadratic equation of the form:
$$ax^2 + bx + c = 0$$

Your function should:
- Take three real numbers $a$, $b$, and $c$ as inputs, with $a$ defaulting to 1.
- Return the real solutions of the equation (if any) in a suitable format.

In [ ]:
# Your code here
import math

def QuadEqu(b, c, a=1):
    # Calcul of delta
    delta = b**2 - 4*a*c
    if delta < 0:
        return "No real solutions"
    elif delta == 0:
        x = -b / (2*a)
        return (x,)
    else:
        sqrt_delta = math.sqrt(delta)
        x1 = (-b + sqrt_delta) / (2*a)
        x2 = (-b - sqrt_delta) / (2*a)
        return (x1, x2)

ii) Test the function `QuadEqu` for:
  - $a=1, b=2, c=1$
  - $a=2, b=2, c=40$
  - $a=2, b=3, c=-5$

In [108]:
# Your code here
print("Test for a=1, b=2, c=1   :", QuadEqu(2,3,1))
print("Test for a=1, b=2, c=40  :", QuadEqu(2,2,40))
print("Test for a=1, b=3, c=-5  :", QuadEqu(2,3,-5))


Test for a=1, b=2, c=1   : No real solutions
Test for a=1, b=2, c=40  : No real solutions
Test for a=1, b=3, c=-5  : (-0.6, 1.0)


c) Write a Python program named `fact2nddeg` that repeatedly offers the user a menu with three options:
- Compute a factorial using factorial2
- Solve a quadratic equation using QuadEqu
- Exit the program

The program should:
- Ask the user to select an option.
- Request the necessary input values for the chosen option.
- Call the corresponding function and display the result.
- Repeat the process until the user chooses to exit.

In [97]:
# Your code here
def fact2nddeg():
    
    while True:
        print("\nMenu:")
        print("1. Compute a factorial using factorial2")
        print("2. Solve a quadratic equation using QuadEqu")
        print("3. Exit")
        choice = input("Select an option (1/2/3): ")

        if choice == '1':
            try:
                n = int(input("Enter a non-negative integer: "))
                result = factorial2(n)
                print("Result:", result)
            except ValueError:
                print("Invalid input. Please enter a valid integer.")
        elif choice == '2':
            try:
                a = float(input("Enter coefficient a: "))
                b = float(input("Enter coefficient b: "))
                c = float(input("Enter coefficient c: "))
                roots = QuadEqu(a, b, c)
                if not roots:
                    print("No real solutions.")
                else:
                    print("Solutions:", roots)
            except ValueError:
                print("Invalid input. Please enter valid numbers.")

        elif choice == '3':
            print("Exiting the program.")
            break

        else:
            print("Invalid option. Please choose 1, 2, or 3.")

if __name__ == "__main__":
    fact2nddeg()


Menu:
1. Compute a factorial using factorial2
2. Solve a quadratic equation using QuadEqu
3. Exit
Result: 479001600

Menu:
1. Compute a factorial using factorial2
2. Solve a quadratic equation using QuadEqu
3. Exit
Exiting the program.


### Exercise 2
This exercise builds the essential mathematical tools used in public-key cryptography, like [**Rivest–Shamir–Adleman (RSA)**](https://en.wikipedia.org/wiki/RSA_cryptosystem). Security relies on the unique properties of prime numbers and the extreme difficulty of factoring very large numbers.

The Core Principle: Prime Factoring is Hard
   - **Easy**: Multiplying two large prime numbers ($p$ and $q$) to get a huge composite number ($N=p \cdot q$) is very fast for computers.
   - **Hard**: Given only the product $N$, finding the original two prime factors, $p$ and $q$, takes an enormous amount of time. The time required grows exponentially with the size of $N$. This difficulty is what guarantees security.

a) Write a function called `number_of_factors` that takes a positive integer $n$ and returns how many factors the number has together with a list of its factors.

In [ ]:
# Your code here
def number_of_factors(n):
    if n <= 0:
        return (0, [])
    factors = []
    for i in range(1, n + 1):
        if n % i == 0:
            factors.append(i)

    return (len(factors), factors)

b) Write a Python function called `is_prime` that takes a positive integer, $n$, as input and returns the boolean True if $n$ is a prime number and False, otherwise.

Note : A prime number (or a prime) is a natural number greater than 1 and that has no positive divisors other than 1 and itself.

In [ ]:
# Your code here
def is_prime(n):
    if n <= 1:
        return False
    if n == 2:
        return True 
    if n % 2 == 0:
        return False

    for i in range(3, int(n**0.5) + 1, 2):
        if n % i == 0:
            return False
    return True

c) Cryptographic systems need a large pool of primes. Write a function called `primes(k)` that returns a list of the first k prime numbers. (Set the default value of $k$ to 100).

In [ ]:
# Your code here

d) Modify the `primes` function to accept an optional argument called start (default value is 2). The function should now return the first $k$ prime numbers that are greater than or equal to start. This simulates searching for primes in a specific high-security range.

In [ ]:
# Your code here


e) The [**Fundamental theorem of arithmetic**](https://en.wikipedia.org/wiki/Fundamental_theorem_of_arithmetic) states that every integer $n$ is a prime number or can be uniquely written as a product of prime powers (e.g., $12=2^2 \cdot 3^1$. We are interested in finding the exponent of a specific prime in such a factorization. Write a function `V(n, p)` that takes a positive integer $n$ and a prime $p$, and returns the largest integer $r$ such that $p^r$ divides $n$. This value, $r$, is known as the [**p-adic valuation**](https://en.wikipedia.org/wiki/P-adic_valuation).

In [110]:
# Your code here
def V(n, p):
    if n <= 0 or p <= 1:
        raise ValueError("n must be > 0 and p must be a prime > 1")
    r = 0
    while n % p == 0:
        n = n // p
        r += 1
    return r

In [ ]:
print(V(12, 2))
print(V(12, 3))
print(V(12, 5))   
print(V(1000, 2)) 
print(V(1000, 5))

2
1
0
3
3


f) We would like now to compare the prime-factor structure of two numbers, $n$ and $m$, that is, we would like to measure how different two numbers are in terms of the powers of the primes in their factorizations. Write a function `D(n, m)` that takes two positive integers, $n$ and $m$, and returns a total score $S_1$ calculated as follows:

 $$ S_1 = \sum_{p\; prime} \Bigl| \nu_p(n) - \nu_p(m)\Bigr| $$

Note that `D(14,24)=4`.



In [ ]:
# Your code here



g) The final problem requires you to analyze the average behavior of the `D(n,m)` function by computing a large, iterated sum.  Hence, write a function `P` that takes a positive integer `N` and returns the sum $S_2$ given by:

$$S_2  = \sum_{n=1}^{N}\sum_{m=1}^{N}D(n,m).$$

Note that $P(10)=210$ and $P(10^2) = 37018$.


In [ ]:
# Your code here

### Exercise 3

Imagine a population of bacteria that grows at a rate proportional to its current size, but with limited food supply. Suppose its growth is modeled by

$$\dfrac{dy}{dt}=0.5y$$

with initial population $y(0)=2$.

a) Explain in words what this equation means about how the population changes over time.

**Your explanation here**

b) [**Euler’s method**](https://en.wikipedia.org/wiki/Euler_method) is a numerical technique to approximate solutions of ordinary differential equations (ODEs) when an exact solution is hard (or impossible) to find.

Suppose we have an ODE:
$$\dfrac{dy}{dt} = f(t, y), \quad y(t_0)=y_0$$

The idea is:

- Start from an initial point $(t_0, y_0)$
- Move forward in small steps of size $h$
- At each step, estimate the new value of $y$ using the slope $f(t,y)$ at the current point

The update formula is thus given by:
$$y_{n+1} = y_n + h \cdot f(t_n,y_n)$$
$$t_{n+1} = t_n + h$$

The pseudocode is given as follows:

c) i) Write $f(t, y) = 0.5y$ as a lambda function.

ii) Now, use Euler’s method with a step size $h=0.1$ to approximate the population at $t=0.3$. Strictly use lists.

d) Compare your result at $t=0.3$ with the exact solution $y(t) = 2\exp^{0.5t}$, that is, compute the absolute error.

e) Compare your answers for $y(0.3)$ with $h=0.1$ and $h=0.05$. Which one is closer to the exact solution $y(0.3) = 2\exp^{0.15}$ ? Reflect on this.

### Exercise 4

Determine how many Sundays fell on the first day of a month during the twentieth century, that is, from 1 January 1901 to 31 December 2000.

The following information might be useful (though you may also verify or research independently):

- 1 January 1900 was a Monday.
- The number of days in each month is as follows:
    - September, April, June, and November each have 30 days.
    - All other months have 31 days, except February, which has 28 days in a common year and 29 days in a leap year.
- Leap year rules:
    - A year is a leap year if it is divisible by 4.
    - However, if the year is a century (i.e., divisible by 100), it is only a leap year if it is also divisible by 400.

171
